### DISCUSSION TITLE
请问作者有适配sing-box的想法吗

### DISCUSSION THREAD

**Discussion by yyiwu** [lxhao61/integrated-examples#168] [Q&A] [answered]
我个人用的是sing-box因为它有tun模式并且自带h2mux，但是sing-box没有回落我不知道如何去按照您的xray的配置去把sing-box的配置贴进去

**Comment by lxhao61** [comment, marked as answer]
没有。个人觉得sing-box就是有tun模式、自带h2mux、无回落等鲜明特点，最合适做第三方客户端、不适合做组合应用服务端。而本项目主要是分享怎么搭建主流科学上网的优化配置及最优组合示例，觉得无必要折腾这个服务端应用。
另外你实在要用sing-box做服务端，那么与回落相关应用只有删除，即以我的服务端示例做减法应用。

**Reply by choicky** [reply]
我前几个月折腾了一下，把服务器端的 `xray` 换成了 `sing-box` 了。
sing-box 支持的协议比较多，vless, hysteria2, ss 都可以，一把梭，方便一些。

**Comment by choicky** [comment]
我服务器就是参考 @lxhao61 大佬的配置，用 `xray` 再跑。
本地端最近在使用 `sing-box` 。主要是我有些软件不受 `v2rayN` 管，不走 `v2rayN` 。`sing-box` 的 `tun` 能解决这个问题。

**Comment by bash99** [comment]
sing-box服务端只支持 xless + vision + reality和torjan的回落。

sing-box和mihomo似乎思路一致，协议支持比较一致，比如都引入anytls，都没支持xhttp（而且号称不会支持了）。
它们（包括anytls的作者）似乎都认为fallback不太重要（比如sing-box服务端唯一支持fallback的trojan协议的部分文档），反而更重视头几个包的scheme识别风险。

而xray的作者显然认为回落很重要，同时认为ss-2022-aead已经极端危险。

当然从国内现实来看，暂时确实没fallback这些协议没被封，甚至ss-2022-aead直接过墙也不会被封，因此这个争论可能只有下一次大规模封锁才能见分晓了。

我也有两个服务器是采用lxhao61大佬的配置，但是caddy我实在更喜欢Caddyfile格式而不是json格式，因此只follow到xless + vision + reality阶段。

**Reply by hide3110** [reply]
能参考一下大佬singbox reality用caddy分流sni复用443的配置文件吗？我用nginx前置分流sni速度影响还挺大，感觉效率降低了。

**Reply by choicky** [reply]
之前我和大佬 @lxhao61 做过测试， caddy sni 分流性能比 nginx 强太多了。 Ref: https://github.com/lxhao61/integrated-examples/discussions/186

**Reply by choicky** [reply]
> 能参考一下大佬singbox reality用caddy分流sni复用443的配置文件吗？我用nginx前置分流sni速度影响还挺大，感觉效率降低了。

基于大佬的前置[Xray(M+F+H+K+A)+Nginx](https://github.com/lxhao61/integrated-examples/tree/new/Xray(M%2BF%2BH%2BK%2BA)%2BNginx)（由 Nginx 兼顾 SNI 分流的综合应用。） 微调一下就可以。

例如，  caddy 前置 + xray 后端， caddy 的配置是 
```
            "handle": [{
              "handler": "proxy",
              "proxy_protocol": "v1", //开启 PROXY protocol 发送。v1 或 v2 表示 PROXY protocol 版本。
              "upstreams": [{
                "dial": ["127.0.0.1:5443"] //转给 VLESS+Vision+REALITY 本地监听端口
              }]
            }]
```

你删掉 `proxy_protocol` 那一行，即可使用 sing-box 代替 xray 了。

**Reply by hide3110** [reply]
是参考哪份caddy的配置删proxy_protocol啊？另外，caddy前置速度会回来吗？对比reality高段位端口的速度差别不大吧？用nginx前置sni分流后速度下降厉害。
想试试singbox前置+caddy，但不知道怎么配？

**Reply by lxhao61** [reply]
应该写错了，是让你参考 [Xray(M+F+H+K+A)+Caddy(N)](https://github.com/lxhao61/integrated-examples/tree/new/Xray(M%2BF%2BH%2BK%2BA)%2BCaddy(N)) 示例修改。
另外 VPS 性能不好且要速度、建议还是不要 SNI 分流（Caddy 或 Nginx 前置），推荐选 Xray 前置示例。

**Reply by hide3110** [reply]
确实，nginx和caddy前置试了个遍，速度都下降不少，sing-box reality还没有回落，没得玩，难怪没人分享与sing-box组合的配置，只剩照抄被群主玩花了的xray就行了。

**Reply by bash99** [reply]
> sing-box reality还没有回落

sing-box reality 靠偷自己勉强也能回落？

**Reply by choicky** [reply]
> sing-box reality 靠偷自己勉强也能回落？

印象中，sing-box 不喜欢回落，没有提供这个功能。

反正我没有搞定回落；如大佬们搞定了，分享一下。

不过，我觉得…… 有回落，可能只是心理安慰吧

**Reply by bash99** [reply]
> 印象中，sing-box 不喜欢回落，没有提供这个功能。
> 
> 反正我没有搞定回落；如大佬们搞定了，分享一下。
> 
> 不过，我觉得…… 有回落，可能只是心理安慰吧

大概这么写，当然这里我caddy跑在443端口，完整搞法应该是caddy配置改成https监听非443端口，然后vless跑在443
```
    {
      "type": "vless",
      "listen": "::",
      "listen_port": 16443,
      "users": [
        {
          "uuid": "xuuid",
          "flow": "xtls-rprx-vision"
        }
      ],
      "tls": {
        "enabled": true,
        "server_name": "myserver",
        "reality": {
          "enabled": true,
          "handshake": {
            "server": "myserver",
            "server_port": 443
          },
          "private_key": "prk",
          "short_id": ["sid"]
        }
      }
    }
```

**Comment by choicky** [comment]
我的 `caddy` 监听正常的80、443以及8443端口。 `sing-box` 监听7443端口。
`caddy` 通过 `sni` 把域名 `vlessr.example.com` 的443转发到 `sing-box` 的 7443端口，把域名 `www.example.com` 的443转发到`caddy` 的 8443端口 。

目前基本正常，但域名 `vlessr.example.com` 无法回落到 8443。大伙们能使用 sing-box 回落吗？

`sing-box` 目前是这样
```
{
  //其他略
  "inbounds": [
    {
      "type": "vless",
      "tag": "reality-xtls-in",
      "listen": "::",
      "listen_port": 7443,
      "users": [
        {
          "name": "某email",
          "uuid": "某UUID",
          "flow": "xtls-rprx-vision"
        }
      ],
      "tls": {
        "enabled": true,
        "server_name": "vlessr.example.com",
        "reality": {
          "enabled": true,
          "handshake": {
            "server": "vlessr.example.com",
            "server_port": 8443
          },
          "private_key": "某KEY",
          "short_id": [
            ""
          ]
        }
      }
    }
  ]
}
```

`caddy` 大概是这样
```
{
  "admin": {
    "disabled": true, //禁用 admin API。（选配）
    "config": {
      "persist": false
    }
  },
  "logging": {
    "logs": {
      "default": {
        "writer": {
          "output": "file",
          "filename": "/var/log/caddy/error.log" //错误日志的文件地址
        },
        "encoder": {
          "format": "console"
        },
        "level": "ERROR",
        "exclude": [
          "http.log.access.log0" //启用访问日志需要的配置，否则必须删除。
        ]
      },
      "log0": {
        "writer": {
          "output": "file",
          "filename": "/var/log/caddy/access.log" //访问日志的文件地址
        },
        "encoder": {
          "format": "console"
        },
        "include": [
          "http.log.access.log0"
        ]
      } //访问日志的输出配置。（选配）
    }
  },
  "storage": {
    "module": "file_system",
    "root": "/home/tls" //存放 TLS 证书的基本路径
  },
  "apps": {
    "layer4": {
      "servers": {
        "tcpsni": {
          "listen": [
            ":443"
          ],
          "routes": [
            {
              "match": [
                {
                  "tls": {
                    "sni": [ //普通网站的域名
                      "www.example.com"
                    ]
                  }
                }
              ],
              "handle": [
                {
                  "handler": "proxy",
                  "proxy_protocol": "v2", //开启 PROXY protocol 发送。v1 或 v2 表示 PROXY protocol 版本。
                  "upstreams": [
                    {
                      "dial": [
                        "127.0.0.1:8443" //转给 HTTP/3 server 本地监听端口，即 www.example.com 的真实监听地址
                      ]
                    }
                  ]
                }
              ]
            },
            {
              "match": [
                {
                  "tls": {
                    "sni": [
                      "vlessr.example.com" //对应 VLESS+Vision+REALITY 的域名
                    ]
                  }
                }
              ],
              "handle": [
                {
                  "handler": "proxy", //后端为sing-box所以不再附加 "proxy_protocol": "v1" ，v1 或 v2 表示 PROXY protocol 版本。
                  "upstreams": [
                    {
                      "dial": [
                        "127.0.0.1:7443" //转给 VLESS+Vision+REALITY 本地监听端口
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    },
    "http": {
      "servers": {
		   //略
	  }
	},
    "tls": {
      "certificates": {
        "automate": [ //自动化管理 TLS 证书（包括获取与更新证书）
          "www.example.com",
          "vlessr.example.com"
        ]
      },
      "automation": {
        "policies": [
          {
            "issuers": [
              {
                "module": "acme",
                "email": "iLoveChina@example.com"
              }
            ]
          }
        ]
      }
    }
  }
}


```

**Reply by lxhao61** [reply]
sing-box 中 "server": "vlessr.example.com" 改为 "server": "127.0.0.1" 试试。
另外取消 8443 端口上 PROXY protocol 发与收配置。

**Reply by choicky** [reply]
我记混了。我的配置中， vlessr.example.com 的回落是正常的。

我还有另一个 inbound，是传统的 E=VLESS+Vision+TLS 模式。 xray 通过 fallbacks 回落到一个 h2c 的web 。 sing-box 好像不支持回落，所以我使用 sing-box 跑服务时， 没办法设置回落。不知道大佬是否有什么解决方案。

Ref: 
https://github.com/lxhao61/integrated-examples/blob/main/Xray(E%2BF%2BH%2BA)%2BCaddy(N)/1_xray.jsonc

**Reply by lxhao61** [reply]
没办法，故仅推荐使用 Xray 为服务端。
